### ISSUE TITLE
`psc-make` runs out of memory rather than returning proper error

### ISSUE THREAD

**Issue by pseudonom** [purescript/purescript#979]
With this code:

```purescript
module Foo where

import qualified Data.Map as M
import Data.Tuple

norm :: forall a. (Ord a) => [Tuple a Number] -> [Tuple a Number]
norm = M.toList <<< M.fromListWith (+)

newtype Dist a = Dist [Tuple a Number]

instance semiringDist :: (Semiring a, Ord a) => Semiring (Dist a) where
  (+) x y = Dist []
  (*) x y = Dist []
  one = Dist []
  zero = Dist []
instance moduloSemiringDist :: (ModuloSemiring a, Ord a) =>
         ModuloSemiring a where
  (/) x y = Dist []
  mod x y = Dist []
```

the `moduloSemiring` instance is erroneously for `a` instead of `Dist a`. 

Rather than point out this error, `psc-make` grinds away for awhile and eventually throws `psc-make: out of memory (requested 1048576 bytes)`. 

Strangely, each element in the above code seems necessary for the problem to present.

**Comment by paf31** [comment]
This should be fixed by my changes which terminate the search after a finite number of steps. The problem is that the superclass search for `moduloSemiringDist` will just spin.

**Comment by pseudonom** [comment]
Is that fix in 0.6.9? Because I'm getting the problem in 0.6.9.

**Comment by paf31** [comment]
Nope, not yet.
